common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java

```diff
@@ -423,18 +440,51 @@ protected ServiceInfo resolveByPartitionIdx(List<ServiceInfo> servers, QueueKey
             return null;
         }
 
-        if (!ServiceType.TB_RULE_ENGINE.equals(queueKey.getType()) || TenantId.SYS_TENANT_ID.equals(queueKey.getTenantId())) {
-            return servers.get(partition % servers.size());
-        } else {
-            int hash = hashFunction.newHasher().putLong(queueKey.getTenantId().getId().getMostSignificantBits())
-                    .putLong(queueKey.getTenantId().getId().getLeastSignificantBits())
+        TenantId tenantId = queueKey.getTenantId();
+        if (queueKey.getType() == ServiceType.TB_RULE_ENGINE) {
+            if (!responsibleServices.isEmpty()) { // if there are any dedicated servers
+                TenantProfileId profileId;
+                if (tenantId != null && !tenantId.isSysTenantId()) {
+                    TenantRoutingInfo routingInfo = tenantRoutingInfoService.getRoutingInfo(tenantId);
+                    profileId = routingInfo.getProfileId();
+                } else {
+                    profileId = null;
+                }
+
+                List<ServiceInfo> responsible = responsibleServices.get(profileId);
+                if (responsible == null) {
+                    // if there are no dedicated servers for this tenant profile, or for system queues,
+                    // using the servers that are not responsible for any profile
+                    responsible = servers.stream()
```

Comment:
I don't like the idea of manipulating the ```responsibleServices``` outside the ```recalculateParititions```. You may just create new variable and update it in the ```recalculateParititions```. Somethings like ```commonServices``` - list of services that process data from generic tenant profiles.